diff --git a/dlls/winealsa.drv/alsa.c b/dlls/winealsa.drv/alsa.c
index 285d7ecf6de..c27664bb254 100644
--- a/dlls/winealsa.drv/alsa.c
+++ b/dlls/winealsa.drv/alsa.c
@@ -2521,6 +2521,7 @@ const unixlib_entry_t __wine_unix_call_funcs[] =
     alsa_get_capture_buffer,
     alsa_release_capture_buffer,
     alsa_is_format_supported,
+    alsa_not_implemented,
     alsa_get_mix_format,
     alsa_get_device_period,
     alsa_get_buffer_size,
@@ -2978,6 +2979,7 @@ const unixlib_entry_t __wine_unix_call_wow64_funcs[] =
     alsa_wow64_get_capture_buffer,
     alsa_release_capture_buffer,
     alsa_wow64_is_format_supported,
+    alsa_not_implemented,
     alsa_wow64_get_mix_format,
     alsa_wow64_get_device_period,
     alsa_wow64_get_buffer_size,
