fixes https://bugs.winehq.org/show_bug.cgi?id=56372
diff --git a/libs/musl/Makefile.in b/libs/musl/Makefile.in
index 11111111111..11111111111 100644
--- a/libs/musl/Makefile.in
+++ b/libs/musl/Makefile.in
@@ -1,6 +1,6 @@
 EXTLIB    = libmusl.a
 EXTRAINCL = -I$(srcdir)/src/internal -I$(srcdir)/arch/generic
-EXTRADEFS = -D_ACRTIMP= -D_NO_CRT_MATH_INLINE
+EXTRADEFS = -D_ACRTIMP= -D_NO_CRT_MATH_INLINE $(MUSL_I386_FLAGS)
 
 SOURCES = \
 	src/math/__cos.c \
diff --git a/configure.ac b/configure.ac
index 11111111111..11111111111 100644
--- a/configure.ac
+++ b/configure.ac
@@ -925,6 +925,8 @@ do
             AS_VAR_APPEND([${wine_arch}_EXTRACFLAGS],[" $llvm_extra_cflags"])
             AS_VAR_APPEND([${wine_arch}_LDFLAGS],[" $llvm_extra_ldflags"])
             CFLAGS="$llvm_extra_cflags $llvm_cflags"
+        else
+            AC_SUBST(MUSL_I386_FLAGS,"-ffloat-store")
         fi
         AS_VAR_SET([${wine_arch}_TARGET],[$target])])
     AS_VAR_POPDEF([wine_cv_crosscc])
