fixes https://bugs.winehq.org/show_bug.cgi?id=56372
diff --git a/libs/musl/Makefile.in b/libs/musl/Makefile.in
index 11111111111..11111111111 100644
--- a/libs/musl/Makefile.in
+++ b/libs/musl/Makefile.in
@@ -1,6 +1,6 @@
 EXTLIB    = libmusl.a
 EXTRAINCL = -I$(srcdir)/src/internal -I$(srcdir)/arch/generic
-EXTRADEFS = -D_ACRTIMP= -D_NO_CRT_MATH_INLINE
+EXTRADEFS = -D_ACRTIMP= -D_NO_CRT_MATH_INLINE $(MUSL_I386_FLAGS)
 
 SOURCES = \
 	src/math/__cos.c \
diff --git a/configure.ac b/configure.ac
index 11111111111..11111111111 100644
--- a/configure.ac
+++ b/configure.ac
@@ -921,6 +921,9 @@ do
         AS_VAR_APPEND([${wine_arch}_LDFLAGS],[" $llvm_extra_ldflags"])
         CFLAGS="$llvm_extra_cflags $llvm_cflags"
     else
+        case $wine_arch in
+            i386) AC_SUBST(MUSL_I386_FLAGS,"-ffloat-store") ;;
+        esac
         AS_VAR_PUSHDEF([wine_cv_crosscc], [wine_cv_${wine_arch}_crosscc])
         AC_CACHE_CHECK([whether $CC works], wine_cv_crosscc,
             [AC_COMPILE_IFELSE([AC_LANG_PROGRAM([])], [AS_VAR_SET([wine_cv_crosscc],[yes])], [AS_VAR_SET([wine_cv_crosscc],[no])])])
