diff --git a/Makefile.in b/Makefile.in
index 44d0447e..c215d823 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -85,7 +85,7 @@ all: all-dist | $(DST_DIR)
 all-dist: | $(DST_DIR)
 dist: all
 
-J = $(patsubst -j%,%,$(filter -j%,$(MAKEFLAGS)))
+J = $(patsubst -j%,%,$(SUBJOBS))
 
 include $(SRC)/make/utility.mk
 include $(SRC)/make/rules-source.mk
diff --git a/make/rules-autoconf.mk b/make/rules-autoconf.mk
index 878640d2..614b579f 100644
--- a/make/rules-autoconf.mk
+++ b/make/rules-autoconf.mk
@@ -29,9 +29,9 @@ $$(OBJ)/.$(1)-configure$(3): $$($(2)_SRC)/configure
 $$(OBJ)/.$(1)-build$(3):
 	@echo ":: building $(3)bit $(1)..." >&2
 	+cd "$$($(2)_OBJ$(3))" && env $$($(2)_ENV$(3)) \
-	$$(BEAR) $$(MAKE)
+	$$(BEAR) $$(MAKE) -j$$(SUBJOBS)
 	cd "$$($(2)_OBJ$(3))" && env $$($(2)_ENV$(3)) \
-	$$(MAKE) install
+	$$(MAKE) -j$$(SUBJOBS) install 
 	touch $$@
 endef
 
diff --git a/make/rules-cargo.mk b/make/rules-cargo.mk
index d80ff9ed..e9434078 100644
--- a/make/rules-cargo.mk
+++ b/make/rules-cargo.mk
@@ -14,7 +14,7 @@ $$(OBJ)/.$(1)-build$(3):
 	@echo ":: building $(3)bit $(1)..." >&2
 	cd $$($(2)_SRC) && env $$($(2)_ENV$(3)) \
 	cargo build \
-	      $$(filter -j%,$$(MAKEFLAGS)) \
+	      -j$$(SUBJOBS) \
 	      --target "$$(CARGO_TARGET_$(3))" \
 	      --target-dir $$($(2)_OBJ$(3)) \
 	      $$(CARGO_BUILD_ARGS) \
diff --git a/make/rules-cmake.mk b/make/rules-cmake.mk
index 520d88ac..ab898810 100644
--- a/make/rules-cmake.mk
+++ b/make/rules-cmake.mk
@@ -28,9 +28,9 @@ $$(OBJ)/.$(1)-configure$(3): $$($(2)_SRC)/CMakeLists.txt
 $$(OBJ)/.$(1)-build$(3):
 	@echo ":: building $(3)bit $(1)..." >&2
 	cd "$$($(2)_OBJ$(3))" && env $$($(2)_ENV$(3)) \
-	$$(MAKE)
+	$$(MAKE) -j$$(SUBJOBS)
 	cd "$$($(2)_OBJ$(3))" && env $$($(2)_ENV$(3)) \
-	$$(MAKE) install
+	$$(MAKE) -j$$(SUBJOBS) install
 	touch $$@
 endef
 
diff --git a/make/rules-configure.mk b/make/rules-configure.mk
index a2205c1..be2abd1 100644
--- a/make/rules-configure.mk
+++ b/make/rules-configure.mk
@@ -25,9 +25,9 @@ $$(OBJ)/.$(1)-configure$(3):
 $$(OBJ)/.$(1)-build$(3):
 	@echo ":: building $(3)bit $(1)..." >&2
 	+cd "$$($(2)_OBJ$(3))" && env $$($(2)_ENV$(3)) \
-	$$(BEAR) $$(MAKE)
+	$$(BEAR) $$(MAKE) -j$$(SUBJOBS)
 	cd "$$($(2)_OBJ$(3))" && env $$($(2)_ENV$(3)) \
-	$$(MAKE) install
+	$$(MAKE) -j$$(SUBJOBS) install
 	touch $$@
 endef

diff --git a/make/rules-makedep.mk b/make/rules-makedep.mk
index 140cb8ea..686c8c1c 100644
--- a/make/rules-makedep.mk
+++ b/make/rules-makedep.mk
@@ -45,9 +45,9 @@ $$(OBJ)/.$(1)-configure$(3):
 $$(OBJ)/.$(1)-build$(3):
 	@echo ":: building $(3)bit $(1)..." >&2
 	+cd "$$($(2)_OBJ$(3))" && env $$($(2)_ENV$(3)) \
-	$$(BEAR) $$(MAKE)
+	$$(BEAR) $$(MAKE) -j$$(SUBJOBS)
 	cd "$$($(2)_OBJ$(3))" && env $$($(2)_ENV$(3)) \
-	$$(MAKE) install
+	$$(MAKE) -j$$(SUBJOBS) install
 	touch $$@
 endef
 
diff --git a/make/rules-meson.mk b/make/rules-meson.mk
index 985f2085..f4f32d94 100644
--- a/make/rules-meson.mk
+++ b/make/rules-meson.mk
@@ -61,7 +61,7 @@ $$(OBJ)/.$(1)-configure$(3): $$($(2)_SRC)/meson.build
 $$(OBJ)/.$(1)-build$(3):
 	@echo ":: building $(3)bit $(1)..." >&2
 	+env $$($(2)_ENV$(3)) \
-	ninja -C "$$($(2)_OBJ$(3))" install
+	ninja -j$$(SUBJOBS) -C "$$($(2)_OBJ$(3))" install
 	touch $$@
 endef
 
diff --git a/make/rules-winemaker.mk b/make/rules-winemaker.mk
index 5c3b59a3..ed6589e0 100644
--- a/make/rules-winemaker.mk
+++ b/make/rules-winemaker.mk
@@ -28,7 +28,7 @@ $$(OBJ)/.$(1)-build$(3):
 	@echo ":: building $(3)bit $(1)..." >&2
 	rsync -arx "$$($(2)_SRC)/" "$$($(2)_OBJ$(3))/"
 	env $$($(2)_ENV$(3)) \
-	$$(MAKE) -C "$$($(2)_OBJ$(3))" LIBRARIES="$$($(2)_LDFLAGS)"
+	$$(MAKE) -j$$(SUBJOBS) -C "$$($(2)_OBJ$(3))" LIBRARIES="$$($(2)_LDFLAGS)"
 	cd "$$($(2)_OBJ$(3))" && touch "$(basename $(4)).spec" && env $$($(2)_ENV$(3)) \
 	winebuild --$(lastword $(subst ., ,$(4))) --fake-module -E "$(basename $(4)).spec" -o "$(4).fake"
 	mkdir -p $$($(2)_LIBDIR$(3))/$(LIBDIR_WINE_$(3))
