diff --git a/Makefile.in b/Makefile.in
index c01116e5..4e5936c5 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -41,7 +41,7 @@ DST_LIBDIR64 := $(DST_DIR)/lib64
 DIST_PREFIX := $(DST_DIR)/share/default_pfx/
 DIST_VERSION := $(DST_BASE)/version
 DEPLOY_DIR := ./deploy
-REDIST_DIR := ./redist
+REDIST_DIR := ./$(BUILD_NAME)
 
 ifneq ($(UNSTRIPPED_BUILD),)
     STRIP :=
@@ -545,6 +545,24 @@ $(DIST_WINEOPENXR_JSON64): $(WINEOPENXR_SRC)/wineopenxr64.json default_pfx
 
 all-dist: $(DIST_WINEOPENXR_JSON64)
 
+##
+## protonfixes
+##
+
+PROTONFIXES_TARGET := $(addprefix $(DST_BASE)/,protonfixes)
+$(PROTONFIXES_TARGET): $(addprefix $(SRCDIR)/,protonfixes)
+
+$(OBJ)/.build-protonfixes:
+	cd $(SRCDIR)/protonfixes && make
+	touch $(@)
+
+$(PROTONFIXES_TARGET): $(OBJ)/.build-protonfixes
+	cd $(SRCDIR)/protonfixes && make install
+	#rm -rf $(PROTONFIXES_TARGET)
+	cp -a $(SRCDIR)/protonfixes/dist/protonfixes $(PROTONFIXES_TARGET)
+	rm -r $(SRCDIR)/protonfixes/dist
+
+all-dist: $(PROTONFIXES_TARGET)
 
 ##
 ## steam.exe
@@ -1450,15 +1468,12 @@ deploy: all
 
 .PHONY: redist
 redist: all
-	mkdir -p $(REDIST_DIR)
-	rsync --delete -arx $(DST_BASE)/ $(REDIST_DIR)
-	cp $(PROTONFIXES_TARGET)/cabextract $(REDIST_DIR)/files/bin/
-	cp $(PROTONFIXES_TARGET)/libmspack.so.0 $(REDIST_DIR)/files/lib64/
-	cp $(PROTONFIXES_TARGET)/libmspack.so.0.1.0 $(REDIST_DIR)/files/lib64/
-	mv $(REDIST_DIR) $(BUILD_NAME)
-	tar -cvzf $(BUILD_NAME).tar.gz $(BUILD_NAME)
-	sha512sum $(BUILD_NAME).tar.gz > $(BUILD_NAME).sha512sum
-	@echo "Proton build available at $(BUILD_NAME).tar.gz"
+	mkdir -p $(REDIST_DIR)
+	rsync --delete -arx $(DST_BASE)/ $(REDIST_DIR)
+	cp $(PROTONFIXES_TARGET)/cabextract $(REDIST_DIR)/files/bin/
+	cp -a $(PROTONFIXES_TARGET)/libmspack.so $(REDIST_DIR)/files/lib64/
+	cp -a $(PROTONFIXES_TARGET)/libmspack.so.0 $(REDIST_DIR)/files/lib64/
+	cp $(PROTONFIXES_TARGET)/libmspack.so.0.1.0 $(REDIST_DIR)/files/lib64/
 
 ##
 ## make module=$dllname module[32,64]

i spent 2 hours on realizing this was missing
diff --git a/Makefile.in b/Makefile.in
index d863ac4..49da93b 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -1407,12 +1407,9 @@ $(PROTON37_TRACKED_FILES_TARGET): $(addprefix $(SRCDIR)/,proton_3.7_tracked_file
 USER_SETTINGS_PY_TARGET := $(addprefix $(DST_BASE)/,user_settings.sample.py)
 $(USER_SETTINGS_PY_TARGET): $(addprefix $(SRCDIR)/,user_settings.sample.py)
 
-PROTONFIXES_TARGET := $(addprefix $(DST_BASE)/,protonfixes)
-$(PROTONFIXES_TARGET): $(addprefix $(SRCDIR)/,protonfixes)
-
 DIST_COPY_TARGETS := $(FILELOCK_TARGET) $(PROTON_PY_TARGET) \
-                     $(PROTON37_TRACKED_FILES_TARGET) $(USER_SETTINGS_PY_TARGET) \
-                     $(PROTONFIXES_TARGET)
+                     $(PROTON37_TRACKED_FILES_TARGET) $(USER_SETTINGS_PY_TARGET)
+                     
 
 $(DIST_COPY_TARGETS): | $(DST_DIR)
 	cp -a $(SRCDIR)/$(notdir $@) $@
