diff --git a/Makefile.in b/Makefile.in
index 1551dcb..1ee8c7c 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -41,7 +41,7 @@ DST_LIBDIR64 := $(DST_DIR)/lib64
 DIST_PREFIX := $(DST_DIR)/share/default_pfx/
 DIST_VERSION := $(DST_BASE)/version
 DEPLOY_DIR := ./deploy
-REDIST_DIR := ./redist
+REDIST_DIR := ./$(BUILD_NAME)
 
 ifneq ($(UNSTRIPPED_BUILD),)
     STRIP :=
@@ -1464,7 +1464,6 @@ redist: all
 	cp -a $(PROTONFIXES_TARGET)/libmspack.so $(REDIST_DIR)/files/lib64/
 	cp -a $(PROTONFIXES_TARGET)/libmspack.so.0 $(REDIST_DIR)/files/lib64/
 	cp $(PROTONFIXES_TARGET)/libmspack.so.0.1.0 $(REDIST_DIR)/files/lib64/
-	mv $(REDIST_DIR) $(BUILD_NAME)
 
 ##
 ## make module=$dllname module[32,64]
