From e2003d21cf32c21c475291802d0af1fd385cb0d6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Wed, 29 Nov 2023 23:20:41 +0300
Subject: [PATCH 3/3] server: Check wineserver privileges on init with -20
 niceness.

---
 server/thread.c | 9 +--------
 1 file changed, 1 insertion(+), 8 deletions(-)

diff --git b/server/thread.c a/server/thread.c
index 08742786b03..b0617eda41d 100644
--- b/server/thread.c
+++ a/server/thread.c
@@ -228,14 +228,7 @@ void init_threading(void)
 #ifdef __linux__
 #ifdef RLIMIT_NICE
     struct rlimit rlimit;
-#endif
-#ifdef HAVE_SETPRIORITY
-    /* if wineserver has cap_sys_nice we are unlimited, but leave -20 to the user */
-    if (!setpriority( PRIO_PROCESS, getpid(), -20 )) nice_limit = -19;
-    setpriority( PRIO_PROCESS, getpid(), 0 );
-#endif
-#ifdef RLIMIT_NICE
-    if (!nice_limit && !getrlimit( RLIMIT_NICE, &rlimit ))
+    if (!getrlimit( RLIMIT_NICE, &rlimit ))
     {
         rlimit.rlim_cur = rlimit.rlim_max;
         setrlimit( RLIMIT_NICE, &rlimit );
-- 
2.46.2

