diff --git a/dlls/avrt/main.c b/dlls/avrt/main.c
index 11111111111..11111111111 100644
--- a/dlls/avrt/main.c
+++ b/dlls/avrt/main.c
@@ -56,8 +56,6 @@ HANDLE WINAPI AvSetMmThreadCharacteristicsA(const char *name, DWORD *index)
 
 HANDLE WINAPI AvSetMmThreadCharacteristicsW(const WCHAR *name, DWORD *index)
 {
-    FIXME("(%s,%p): stub\n", debugstr_w(name), index);
-
     if (!name)
     {
         SetLastError(ERROR_INVALID_TASK_NAME);
@@ -70,6 +68,11 @@ HANDLE WINAPI AvSetMmThreadCharacteristicsW(const WCHAR *name, DWORD *index)
         return NULL;
     }
 
+    if (!wcscmp(name, L"Audio") || !wcscmp(name, L"Pro Audio")) {
+        SetThreadPriority(GetCurrentThread(), THREAD_PRIORITY_TIME_CRITICAL);
+        SetThreadDescription(GetCurrentThread(), L"audio_client_avrt");
+    }
+
     return (HANDLE)0x12345678;
 }
 
